### ISSUE TITLE
Support spinning / spinlocks / busy-loops

### ISSUE THREAD

**Issue by sporksmith** [shadow/shadow#1792]
This is a tracking issue for supporting "spin" waiting with `sched_yield`. Today it doesn't work at all - executing a `sched_yield` is effectively a no-op inside of a Shadow simulation. If it's used in a loop that's waiting for another thread on the same host to do something (like release a spin lock), it'll never happen.

*Typically* such spin loops are waiting for a condition that will be satisfied by another thread running on the same host - e.g. releasing a spin lock. It's also possible to end up in a spin loop that's doing non-blocking IO calls, but that's usually a bug (such as https://github.com/smol-rs/async-io/pull/73)

## Spinlock fallback types

Spinlocks often have some type of fallback to a less aggressive waiting mechanism if the condition they're waiting for isn't met for a while.

* No fallback: Without some fallback, currently under Shadow A spinlock that has no never falls back to something less aggressive currently deadlocks, since other threads never get a chance to run.

* Wallclock fallback: A spinlock that spins for some maximum wallclock duration also currently deadlocks, since `sched_yield` neither moves time forward nor gives other threads a chance to run.

* Iteration fallback: A spinlock that spins some maximum number of *iterations* will eventually work. Arguably a "good" spinlock implementation will have such a limit.

## Potential solutions

* Reschedule thread: Implement `sched_yield` to allow other ready threads to run, but not move time forward: This would minimize the change to "simulation-visible" behavior, and could allow Shadow to handle spin-loops where another thread is ready to satisfy the condition (e.g. do some processing and release a spinlock). This is a little tricky though - it'd require adding some way of scheduling a "run this thread again" event at the same sim-time, but after other alread-scheduled events.

* Nanosleep: Implement `sched_yield` as effectively a `nanosleep` of `n` ns. With a very small value (e.g. 1 ns) this would solve at least as many spin-loops as "reschedule thread", but would be much easier to implement. It'd slightly change simulation results, but since even the cheapest syscalls take ~[20 ns](https://stackoverflow.com/questions/23599074/system-calls-overhead), `n` up to that amount would only make the simulation *more* accurate (well, until that minimum syscall latency gets smaller on future systems). This could also fix some spin loops that are stuck waiting for a wallclock fallback, but with a very small `n` could take many iterations to exit the loop. 
 
  Larger values of `n` would reduce the number of such iterations, but at some point could cause the thread in the simulation to be unrealistically slow. e.g. waiting for a full 1s would let shadow exit most such spin loops in a single iteration, but inside the simulation would make the thread seem "slower" than it would be on a real system.

## Real world examples

libopenblas compiled with thread support makes extensive use of spin loops: https://github.com/shadow/shadow/issues/1788. They fallback out of the loop based on "wallclock time", which on x86 ends up being ~270M CPU cycles (270 ms on a 1 GHz CPU). i.e. implementing a "minimal" wait of ~20 ns would cause such loops to be simulated extremely slowly, to the point of being mistaken for deadlock anyway. Implementing a sleep on the order of 1 ms would make these not horribly slow, but we'd probably want to carefully evaluate whether this skews simulation results. 

## Conclusion

In the only current real-world example (libopenblas), we would need `sched_yield` to sleep for a fairly substantial amount of time to avoid substantial overhead in Shadow emulating many iterations of the loop. Since there other workarounds (don't use libopenblas with thread support compiled in, or set `OPENBLAS_NUM_THREADS=1`), it doesn't seem worth potentially skewing other simulation results.

For now, let's keep this issue open and collect other real-world examples as they come up.

**Comment by stevenengler** [comment]
Another real-world example is in older curl versions (such as the version on debian 10), which has an obscure spin loop where time never advances and shadow gets stuck. It does not use sched_yield though, so maybe we should open a different issue for non-sched-yield-based spinning.

**Comment by sporksmith** [comment]
Ah, interesting. I think we can expand the scope of this issue beyond sched_yield; the principle is mostly the same.

Having *all* syscalls (and rdtsc-based-time-checks) move forward by some non-zero amount (instead of just `sched_yield`) could address it. There might be performance impliciations when many threads on the same host are "ready" and executing non-blocking-syscalls though, since we'd end up switching between them after every syscall instead of running one until it hits a blocking syscall.

**Comment by sporksmith** [comment]
@stevenengler IIUC to "fix" #1794 without changing curl, time would need to advance by > 3 ms? Would that have to be in a single iteration of the loop, or would cumulatively over several iterations also work? e.g. if checking the time always advance by 1 microsecond, would we exit the loop after 3k iterations?

**Comment by stevenengler** [comment]
> @stevenengler IIUC to "fix" #1794 without changing curl, time would need to advance by > 3 ms? Would that have to be in a single iteration of the loop, or would cumulatively over several iterations also work? e.g. if checking the time always advance by 1 microsecond, would we exit the loop after 3k iterations?

I think several iterations should work in this case.

> Having all syscalls (and rdtsc-based-time-checks) move forward by some non-zero amount (instead of just sched_yield) could address it. There might be performance impliciations when many threads on the same host are "ready" and executing non-blocking-syscalls though, since we'd end up switching between them after every syscall instead of running one until it hits a blocking syscall.

This would require running the scheduler after every syscall right? Maybe instead we could only increase the cached time in the shim on every syscall, and then only if the new time is greater than the current round end time we treat it as a nanosleep and go back to the scheduler.

**Comment by sporksmith** [comment]
> This would require running the scheduler after every syscall right? Maybe instead we could only increase the cached time in the shim on every syscall, and then only if the new time is greater than the current round end time we treat it as a nanosleep and go back to the scheduler. This wouldn't be ideal when one thread is spinning waiting for another thread, but it should still switch over to the other thread eventually at the end of the current round, and I don't think it would be too bad in the general case when we aren't spinning.

Good point that we already know having to return control from the shim to shadow on every time syscall would be a performance hit (since adding the shim-side-only processing was a win).

The idea of trying to allow the shim to move time forward, with some max upper bound specified by Shadow (such as the end of the current scheduler round, or potentially sooner) could work.

That's getting a bit complex, though. I think waiting for more data points for now is still the right move.

**Comment by sporksmith** [comment]
Thinking about this more after stumbling across the openblas deadlock again...

> Maybe instead we could only increase the cached time in the shim on every syscall, and then only if the new time is greater than the current round end time we treat it as a nanosleep and go back to the scheduler. 

I think this is still my favorite approach, and can be done a little more cleanly now that the time is in shared memory  (instead of sent from shadow to shim every time control is transferred).

Another possibility might be to move time forward and reschedule after X consecutive unblocked syscalls (e.g. 1ms after ~100 syscalls?). This is potentially more efficient since we'd avoid too many context switches in the case where multiple threads on a host aren't blocked, though maybe slightly awkward in terms of the actual simulation semantics.

**Comment by sporksmith** [comment]
After some discussion, the current thinking is to try adding the counter for # of consecutive syscalls, and log a warning when it passes some threshold. We can then use experience with that to think about how to tune the parameters (# of syscalls to trip a warning, how far forward to move time forward).

Both of these parameters could be tunable, and maybe the default for "amount of time to move forward" should be 0, so that it is something the user would explicitly opt into to work around a spin loop (that perhaps they were made aware of in the first place by the warning)

**Comment by robgjansen** [comment]
> maybe the default for "amount of time to move forward" should be 0, so that it is something the user would explicitly opt into to work around a spin loop

I really like this - it provides a convenient way to work around problems, but only in the case that a human confirms that there is indeed a problem :)

**Comment by sporksmith** [comment]
This issue can now be worked around using the experimental `--unblocked-syscall-latency` and `--unblocked-syscall-limit` flags. We're still evaluating and iterating on this feature, though.

**Comment by ppopth** [comment]
I have a real-world example that neither has a non-blocking syscall nor has a fallback mechanism.

```
(gdb) x/10i $rip
=> 0x555556921610 <_ZN4spin4once13Once$LT$T$GT$9call_once17h4de1abff324c11d0E+176>:	pause  
   0x555556921612 <_ZN4spin4once13Once$LT$T$GT$9call_once17h4de1abff324c11d0E+178>:	mov    (%rcx),%rax
   0x555556921615 <_ZN4spin4once13Once$LT$T$GT$9call_once17h4de1abff324c11d0E+181>:	cmp    $0x1,%rax
   0x555556921619 <_ZN4spin4once13Once$LT$T$GT$9call_once17h4de1abff324c11d0E+185>:	
    je     0x555556921610 <_ZN4spin4once13Once$LT$T$GT$9call_once17h4de1abff324c11d0E+176>
```

This code is from a rust crate called "spin".

Since it gets out of the spin only when the memory changes its value and it doesn't call any non-blocking syscall, I'm not sure if there is a way to block the thread and let other threads change the memory value before resuming the spinning thread.

It seems to me that the kernel has to be patched to make the thread blocked, but I'm not sure if there is any other easier way.

Do you have any idea?

**Comment by sporksmith** [comment]
> This code is from a rust crate called "spin".

Interesting; I was going to suggest that the crate should consider adding a `sched_yield` in its spin loops, but it looks like it's specifically targeting `no_std` environments. Technically they could still make a syscall themselves, e.g. with a raw syscall instruction, but such crates are often meant for embedded environments where there are no syscalls. https://crates.io/crates/spin

I think we could potentially handle such spin loops by scheduling a native timer to periodically interrupt the managed thread, move simulated time forward, and potentially yield control back to shadow. It's not trivial to implement or maintain though, and would sacrifice simulation determinism.

Do you have an example of an application that you would like to use under shadow that uses this spin crate? Maybe it's feasible to modify the application to support using more conventional sync primitives when running under an OS (or under shadow)? e.g. it looks like you could enable the [std feature of the spin crate](https://github.com/mvdnes/spin-rs#feature-flags) to have it yield instead of spinning.

**Comment by ppopth** [comment]
> I think we could potentially handle such spin loops by scheduling a native timer to periodically interrupt the managed thread, move simulated time forward, and potentially yield control back to shadow. It's not trivial to implement or maintain though, and would sacrifice simulation determinism.

I thought about interrupting the managed thread as well, but I don't know how to do it yet. Do you have an idea? I can implement it myself for now if I consider that it's worth my effort.

I'm thinking that `ptrace` can do that but I'm also thinking that it's probably an overkill. It will great if there are other ways.

> Do you have an example of an application that you would like to use under shadow that uses this spin crate?

The problem is from running [lighthouse](https://github.com/sigp/lighthouse). The "spin" crate is a dependency of dependencies of lighthouse. I solved the problem by overriding the "spin" crate by following https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html

Thank you so much

**Comment by datacompboy** [comment]
I found another real-world case with pause-based spinloop: MySQL-server.

```
> 164b930:       8b 00                   mov    (%rax),%eax
> 164b932:       83 f8 01                cmp    $0x1,%eax
> 164b935:       75 29                   jne    164b960 <_ZN17Ha_innopart_share20increment_ref_countsEv@@Base+0x320>
> 164b937:       f3 90                   pause  
> 164b939:       48 8b 43 58             mov    0x58(%rbx),%rax
> 164b93d:       48 85 c0                test   %rax,%rax
> 164b940:       75 ee                   jne    164b930 <_ZN17Ha_innopart_share20increment_ref_countsEv@@Base+0x2f0>


 16ca160:       8b 00                   mov    (%rax),%eax
 16ca162:       83 f8 01                cmp    $0x1,%eax
 16ca165:       75 29                   jne    16ca190 <_Z17log_position_lockR5log_t@@Base+0xa0>
 16ca167:       f3 90                   pause  
 16ca169:       48 8b 43 30             mov    0x30(%rbx),%rax
 16ca16d:       48 85 c0                test   %rax,%rax
 16ca170:       75 ee                   jne    16ca160 <_Z17log_position_lockR5log_t@@Base+0x70>
```

There are multiple different busyloops, some have tweakable parameters on number of loops before break, but few doesn't. Basically, one can't just start server.

As workaround for a while I just recompile server with:

```
--- mysql-8.0-8.0.37.orig/storage/innobase/include/ut0ut.h
+++ mysql-8.0-8.0.37/storage/innobase/include/ut0ut.h
@@ -84,7 +84,7 @@ this program; if not, write to the Free
 instruction has important side-effects and must not be removed.
 Also asm volatile may trigger a memory barrier (spilling all registers
 to memory). */
-#define UT_RELAX_CPU() __asm__ __volatile__("pause")
+#define UT_RELAX_CPU() std::this_thread::yield()
 
 #elif defined(HAVE_FAKE_PAUSE_INSTRUCTION)
 #define UT_RELAX_CPU() __asm__ __volatile__("rep; nop")
```

For the "Right" support for busy waits, though, I think about option to:

 - Set real alarm (not emulated) for the thread for 1s (but small enough to not wait forever, large enough to not affect performance of long computations, and large enough to ensure it's "wait forever" to not impact determinism);
 - In case real SIGALRM caught, run step-thru for 30-50 instructions, checking every time is it pause instruction or not (F3 90).
 - If pause instruction found, replace 90 with CC (make it F3 CC).
 - Whenever debug interrupt happened, check is it "F3 CC" or not, if so, treat like yield syscall.

I worry that constant real alarm timeout reset to 1s on each scheduling will impact performance, so, perhaps, it should be per-process setting, not general one; and it seems that means ptrace is mandatory to step-thru and read/write commands memory.

**Comment by sporksmith** [comment]
@datacompboy thank you for the detailed report!

Stepping through and doing dynamic rewriting is a cool idea, but a bit complex, depends on ptrace (breaking compatibility with some environments and simultaneous debugging with gdb etc), and only handles the case that there's a `pause` instruction in the loop.

Another option is to set a relatively short alarm and directly yield in the signal handler. This sacrifices determinism but I think could have ok performance, and I *think* wouldn't be too difficult to implement or maintain. In such cases we could log a warning that determinism has been lost and the user should consider modifying or configuring the software to avoid the busy loop.

In the short term we should definitely add this mysql example and workaround to our documentation.

**Comment by datacompboy** [comment]
While I agree here, partially, I can argue that short alarm will also tamper with simultaneous debugging...

Compatibility with non-pause-hinted spins can be expanded by tracking the loops (i.e. we can not only look for pause instruction but for whole loop cycle) and trying to build spinloop map...

Or static code rewrite 

Or, that may be considered as general limitation of the approach, and documented for good.

**Comment by sporksmith** [comment]
> While I agree here, partially, I can argue that short alarm will also tamper with simultaneous debugging...

I *think* it wouldn't be too bad. If `gdb` is attached, by default it'll stop every time a signal is received, but you can override this behavior and tell it to keep executing. (I think the command is something like `handle SIGALRM noprint`). Conversely, if shadow is `ptrace`-attached to a process, `gdb` won't be able to attach at all.

> Compatibility with non-pause-hinted spins can be expanded by tracking the loops (i.e. we can not only look for pause instruction but for whole loop cycle) and trying to build spinloop map...
> Or static code rewrite
> Or, that may be considered as general limitation of the approach, and documented for good.

Yeah, this approach could be expanded to handle more kinds of loops, but it's a bit of a rabbit hole. It's definitely a cool idea and something to keep in mind, but I think it won't be worth it for us to implement and maintain unless we find more real world use-cases that it would handle; and maybe then only if they can't be somewhat easily patched to `yield` as you've done here for mysql.

In terms of documentation we currently already document that busy loops without any syscalls are unsupported <https://shadow.github.io/docs/guide/limitations.html?highlight=spin#busy-loops>:

> It still doesn't account for time spent by the CPU executing code, which also means that a busy-loop that makes no syscalls at all can still lead to deadlock. Fortunately such busy loops are rare and are generally agreed upon to be bugs, since they'd also potentially monopolize a CPU indefinitely when run natively.

Maybe that paragraph ought to be updated a bit though; in particular I think this is the first real-world example that does this intentionally, somewhat  weakening our argument that it's "generally agreed upon to be a bug". (OTOH maybe it's worth investigating more whether this mysql code would actually be improved by adding some fallback after a set number of loop iterations or clock cycles).

Also we *do* now also handle spin loops that use `rdtsc` or `rdtscp` for clock-cycle/time bounds. We configure the processor to raise an exception on these instructions so that we can move time forward. (I checked, and afaict there is unfortunately not a way to catch the `pause` instruction in this way)

**Comment by datacompboy** [comment]
> "generally agreed upon to be a bug"

The server hangs here: https://github.com/mysql/mysql-server/blob/527c12ed611f3fe072c3043734319edb2c733099/storage/innobase/os/os0thread.cc#L58
Fun enough, it has conditional yield in case if _WIN32, but not linux.

What I don't get, is why there is ifdef for this location, but not many others, like:
https://github.com/mysql/mysql-server/blob/527c12ed611f3fe072c3043734319edb2c733099/storage/innobase/include/os0once.h#L120

That's why I went with very blunt patch to UT_RELAX_CPU. Maybe you right and it is worth on its own to report problem to mysql, but as long as you have preemptive scheduler, it won't be visible -- and the performance for single-core applications definitely not their focus.

  > (I checked, and afaict there is unfortunately not a way to catch the pause instruction in this way)

Yes :( I've spent a while there too and only found some AMD chips have support for SVM (google "SVM_FEATURE_PAUSE_FILTER"), but nothing of the sorts for intel, nor anything similar for yield on ARMv6k+.

But since "pause"/"yield" is only a "hint" and to CPU core only, with less and less single-core real applications, more and more spinloops built nowadays with actual parallelism with no thread scheduler involved...

Given large and heated discussions like https://www.realworldtech.com/forum/?threadid=189711&curpostid=189752 -- I believe the only way forward for non-modified binaries is to become hypervisor...

**Comment by datacompboy** [comment]
Yeah, if you want 100% determinism, then follow the steps of Antithesis seems to be the way: https://antithesis.com/blog/deterministic_hypervisor/

**Comment by sporksmith** [comment]
@ppopth @datacompboy I've split escaping pure-CPU busy loops into issue #2066. I have a draft PR implementing this at #3520. It's probably going to take me a bit to clean up and merge #3520 but I'd be happy to know whether it solves your use-cases (spin crate and mysql server) if you care to try it.

**Comment by ppopth** [comment]
@sporksmith I managed to avoid using busy loops, so it's not a problem for me anymore. However, I still think it's useful to implement in order to avoid future issues.

**Comment by datacompboy** [comment]
I'll do my best to experiment with it next week, will circle back, thank you!

**Comment by sporksmith** [comment]
I think we can close this if and when we declare `--native-preemption-enabled` (added in #3520) to be non-experimental.

**Comment by datacompboy** [comment]
Oh, sorry, forgot to circle back! The native preemption fixes mysql problem; with chdir support the only patch I need to make it start & work is times() syscall support: https://github.com/shadow/shadow/pull/3367#issuecomment-2230830078

but at least the mysql itself no longer needs patching, awesome!
